Extend constraint manifold tutorial with definition format (#494)
ros-planning/moveit_tutorials: doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.rst

It is compactly represented with a list of roll, pitch, yaw and a list of tolerances for each axis, for example::

 - type: orientation
   frame_id: base_link
   link_name: gripper_link
   orientation: [-3.1415269, -1.57079632, 0] #RPY
   tolerances: [6.28318531, 0.2, 6.28318531]
   weight: 1.0
   
VisibilityConstraint
~~~~~~~~~~~~~~~~~~~~

A ``VisibilityConstraint`` allows to eg. specify a camera should always be able to see the gripper.

How this is achieved is best explained by the `VisibilityConstraint <https://docs.ros.org/melodic/api/moveit_core/html/classkinematic__constraints_1_1VisibilityConstraint.html#details>`_ class documentation.

Such a constraint is compactly defined like this::

 - type: visibility
   target_radius: 0.5
   target_pose:
     frame_id: 'base_link'
     position: [1.2, 3.4, 5.6]
     orientation: [-3.1415269, -1.57079632, 0] #RPY
   cone_sides: 4
   sensor_pose:
     frame_id: 'gripper_cam_link'
     position: [1.2, 3.4, 5.6]
     orientation: [-3.1415269, -1.57079632, 0] #RPY
   max_view_angle: 1.1
   max_range_angle: 0.55
   weight: 1.0

Running the database generator
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Assuming MoveIt itself is already launched (via eg. ``roslaunch X_moveit_config demo.launch``), you can use a launch file similar to `generate_state_database.launch <moveit_planners/ompl/ompl_interface/launch/generate_state_database.launch>`_

Reviewer: There seems to be an issue with this link on travis:
```
./build/html/doc/planning_with_approximated_constraint_manifolds/planning_with_approximated_constraint_manifolds_tutorial.html
  *  internally linking to moveit_planners/ompl/ompl_interface/launch/generate_state_database.launch, which does not exist (line 826)
     <a class="reference external" href="moveit_planners/ompl/ompl_interface/launch/generate_state_database.launch">generate_state_database.launch</a>
htmlproofer 3.15.3 | Error:  HTML-Proofer found 1 failure!
The command "htmlproofer ./build --only-4xx --check-html --file-ignore ./build/html/genindex.html,./build/html/search.html --alt-ignore '/.*/' --url-ignore '#' $URL_SWAP" exited with 1.
```
Maybe it has to be the full link address?

Author: Good spot. Should be fixed now.

Creating the Constraint Database
--------------------------------

Constructing a Constraints database is done with the ``generate_state_database`` executable. 

> This loads constraint definition (in a format explained below) from the ROS parameter server and finally outputs the state database to a given directory. 

Reviewer 2: ```suggestion
This loads the constraint definition (in a format explained below) from the ROS parameter server and outputs the state database to a given directory.
```

Defining constraints

> ^^^^^^^^^^^^^^^^^^^^

Reviewer 2: ```suggestion
^^^^^^^^^^^^^^^^^^^^

```

The ``generate_state_database`` executable reads constraints from ROS parameters on ``/constraints``, in a more compact format that a complete ROS message. 
You can define these in ``rosparam`` to be loaded together in a file, eg. ``X_moveit_config/config/constraints.yaml``::

 path_constraint:
   name: some_constraints
   constraints:
   - type: orientation
     frame_id: world
     # etc, as described below

JointConstraint

> ~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~~~~

```

A ``JointConstraint`` limits the positions a joint can take. There are three ways to compactly specify this. 

1. position + a single tolerance
2. position + lower an upper tolerance
3. upper and lower bound

For example::

 - type: joint
   joint_name: first_joint
   position: 0.1
   tolerance: 0.2
   weight: 1.0
 - type: joint
   joint_name: second_joint
   position: 0.1
   tolerances: [0.1, 0.2]
   weight: 1.0
 - type: joint
   joint_name: third_joint
   bounds: [-0.5, 1.0]
   weight: 1.0

PositionConstraint

> ~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~~~~~~~

```

> A ``PositionConstraint`` constraints the cartesian positions allowed for a (position relative to a) link. 

Reviewer 2: ```suggestion
 A ``PositionConstraint`` constraints the Cartesian positions allowed for a (position relative to a) link. 
```

> ``target_offset`` is that relative position wrt. a link, eg. the tip of the end-effector relative to it's mounting point or other origin definition. 

Reviewer 2: ```suggestion
``target_offset`` is that relative position w.r.t. a link, e.g., the tip of the end-effector relative to its mounting point or other origin definition.
```

This region (boxes only in this compact definition) is compactly defined by specifying the upper and lower bounds along each axis. 

For example::

 - type: position
   frame_id: base_link
   link_name: gripper_link
   target_offset: [0.01, 0.01, 0.01]
   region:
     x: [0, 1.0] # [start, end]
     y: [0, 1.0] # [start, end]
     z: [0, 1.0] # [start, end]
   weight: 1.0
   
OrientationConstraint

> ~~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~~~~~~~~~~

```

> An ``OrientationConstraint`` can be used to keep eg. something grasped mostly upright.

Reviewer 2: ```suggestion
An ``OrientationConstraint`` can be used to keep eg. something upright (or mostly upright with respect to some tolerance).
```

It is compactly represented with a list of roll, pitch, yaw and a list of tolerances for each axis, for example::

 - type: orientation
   frame_id: base_link
   link_name: gripper_link
   orientation: [-3.1415269, -1.57079632, 0] #RPY
   tolerances: [6.28318531, 0.2, 6.28318531]
   weight: 1.0
   
VisibilityConstraint

> ~~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~~~~~~~~~

```

How this is achieved is best explained by the `VisibilityConstraint <https://docs.ros.org/melodic/api/moveit_core/html/classkinematic__constraints_1_1VisibilityConstraint.html#details>`_ class documentation.

Such a constraint is compactly defined like this::

 - type: visibility
   target_radius: 0.5
   target_pose:
     frame_id: 'base_link'
     position: [1.2, 3.4, 5.6]
     orientation: [-3.1415269, -1.57079632, 0] #RPY
   cone_sides: 4
   sensor_pose:
     frame_id: 'gripper_cam_link'
     position: [1.2, 3.4, 5.6]
     orientation: [-3.1415269, -1.57079632, 0] #RPY
   max_view_angle: 1.1
   max_range_angle: 0.55
   weight: 1.0

Running the database generator
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Assuming MoveIt itself is already launched (via eg. ``roslaunch X_moveit_config demo.launch``), you can use a launch file similar to `generate_state_database.launch <moveit_planners/ompl/ompl_interface/launch/generate_state_database.launch>`_

The file with the constraint definitions can be passed to the launch file::

 roslaunch ompl_interface generate_state_database.launch constraints_file:=$(rospack find X_moveit_config)/config/constraints.yaml planning_group:=arm

Internals

> ^^^^^^^^^

Reviewer 2: ```suggestion
^^^^^^^^^

```
